tensorboard/webapp/widgets/line_chart_v2/lib/chart.ts

```diff
@@ -0,0 +1,161 @@
+import {Coordinator} from './coordinator';
+import {
+  DataSeries,
+  DataSeriesMetadataMap,
+  Dimension,
+  Extent,
+} from './internal_types';
+import {
+  ObjectRenderer,
+  RendererType,
+  SvgRenderer,
+  ThreeRenderer,
+} from './renderer';
+import {createScale} from './scale';
+import {ScaleType} from './scale_types';
+import {SeriesLineView} from './series_line_view';
+import {ThreeCoordinator} from './threejs_coordinator';
+
+/**
+ * Workaround for testability.
+ *
+ * Inside TensorBoard unit test, we set up zone.js for Angular components and it patches
+ * asynchronousity. For unknown reasons, `fakeAsync` and `flush` misbehave and make test
+ * authoring more cumbersome.
+ */
+const util = {
+  requestAnimationFrame: self.requestAnimationFrame,
+};
+
+export class MainThreadChart implements Chart {
+  private readonly renderer: ObjectRenderer;
+  private readonly seriesLineView: SeriesLineView;
+  private readonly coordinator: Coordinator;
+  private readonly metadataMap: DataSeriesMetadataMap = {};
+  private readonly callbacks: ChartCallbacks;
+
+  constructor(option: ChartOption) {
+    this.callbacks = option.callbacks;
+
+    switch (option.type) {
+      case RendererType.SVG: {
+        this.coordinator = new Coordinator();
+        this.renderer = new SvgRenderer(option.container);
+        break;
+      }
+      case RendererType.WEBGL: {
+        const coordinator = new ThreeCoordinator();
+        this.coordinator = coordinator;
+        this.renderer = new ThreeRenderer(
+          option.container,
+          coordinator,
+          option.devicePixelRatio
+        );
+        break;
+      }
+    }
+
+    this.seriesLineView = new SeriesLineView({
+      renderer: this.renderer,
+      coordinator: this.coordinator,
+      getMetadataMap: () => this.metadataMap,
+    });
+
+    this.resize(option.domDimension);
+  }
+
+  dispose(): void {}
+
+  setXScaleType(type: ScaleType) {
+    this.coordinator.setXScale(createScale(type));
+  }
+
+  setYScaleType(type: ScaleType) {
+    this.coordinator.setYScale(createScale(type));
+  }
+
+  resize(dim: Dimension) {
+    this.coordinator.setDomContainerRect({x: 0, y: 0, ...dim});
+    this.renderer.onResize({x: 0, y: 0, ...dim});
+    this.seriesLineView.setLayoutRect({
+      ...dim,
+      x: 0,
+      y: 0,
+    });
+    this.scheduleRedraw();
+  }
+
+  updateMetadata(metadataMap: DataSeriesMetadataMap) {
+    let shouldRepaint = false;
+    Object.entries(metadataMap).forEach(([id, metadata]) => {
+      const existing = this.metadataMap[id];
+      if (
+        !existing ||
+        metadata.color !== existing.color ||
+        metadata.visible !== existing.visible ||
+        metadata.opacity !== existing.opacity
+      ) {
+        shouldRepaint = true;
+      }
+
+      this.metadataMap[id] = metadata;
+    });
+    if (shouldRepaint) {
+      this.seriesLineView.markAsPaintDirty();
+    }
+    this.scheduleRedraw();
```

Comment:
I changed everything to repaint. The discussion is unclear in my head; they are about the same and did not really think too deeply about the distinction. Would you recommend changing everything to `redraw` since a "paint" is a method of drawing and is an implementation detail?